pkg/infra/serverlock/remote_cache.go

```diff
@@ -0,0 +1,133 @@
+package serverlock
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/remotecache"
+	"github.com/grafana/grafana/pkg/infra/tracing"
+)
+
+const (
+	lockPrefix  = "server_lock:"
+	lockTimeout = 24 * time.Hour * 30
+)
+
+type serverLockRemoteCache struct {
+	remoteCache *remotecache.RemoteCache
+	tracer      tracing.Tracer
+	log         log.Logger
+}
+
+func (sl *serverLockRemoteCache) AcquireLock(ctx context.Context, serverLock *serverLock) (bool, error) {
+	ctx, span := sl.tracer.Start(ctx, "serverLockRemoteCache.acquireLock")
+	defer span.End()
+
+	lockKey := prefixedLockKey(serverLock.OperationUID)
+
+	serverLock.LastExecution = time.Now().Unix()
+	// Convert serverLock to byte array
+	serverLockBytes, err := json.Marshal(serverLock)
+	if err != nil {
+		return false, err
+	}
+
+	err = sl.remoteCache.Set(ctx, lockKey, serverLockBytes, lockTimeout)
```

Comment:
We don't actually have any guarantees here that we haven't clobbered another process attempting to set the same lock at the same time or been clobbered ourselves, because the remote cache doesn't provide an equivalent semantic to the WHERE condition we use in the database-backed locks.

If 2 processes both execute line 53 at the same time for an available lock, they will both call `AcquireLock` and both will appear to have obtained the lock.

There are mechanisms to resolve this problem, but the approach is different for redis (using [redlock](https://redis.io/docs/manual/patterns/distributed-locks/)), memcached (using [compare-and-swap](https://github.com/memcached/memcached/wiki/Commands#cas)) or SQL (using WHERE).

If we want to be able to rely on this we should consider implementing proper lock support in the remote cache service (so we can have backend-specific implementations) and just passing through to that here instead of implementing the existing database algorithm (which itself could be optimized) on top of remote cache.

Reply:
True, without SETNX and pipelines for redis for example we don't have those guarantees. 
This is not an issue for lock implementations that delete themselves but the current method can only be applied to SQL where there are no key expiry mechanisms builtin.

This will require a somewhat deeper rework but I think I'll do it during a hackathon

(references to pick this up again in the future: https://github.com/go-redsync/redsync/blob/master/redis/goredis/v8/goredis.go)